File: Cpp/Reverse Linked List.cpp
/**
 * Definition of ListNode
 *
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @return: The new head of reversed linked list.
     */
    ListNode *reverse(ListNode *head) {
        // write your code here
        stack<ListNode*> st;
        ListNode *pre = new ListNode(0);
        ListNode *res = pre;
        while (head != NULL) {
            st.push(head);
            head = head->next;
        }
        while (!st.empty()) {
            res->next = st.top();
            st.pop();
            res = res->next;
        }
        res->next = NULL;
        return pre->next;
    }
};

File: Java/Reverse-Linked-List.java
/**
 * Definition for ListNode.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int val) {
 *         this.val = val;
 *         this.next = null;
 *     }
 * }
 */ 
public class Solution {
    /**
     * @param head: The head of linked list.
     * @return: The new head of reversed linked list.
     */
    public ListNode reverse(ListNode head) {
        // write your code here
        ListNode cur = null;
        ListNode cnt = head;
        while (cnt != null) {
        	ListNode temp = cnt.next;
        	cnt.next = cur;
        	cur = cnt;
        	cnt = temp;
        }
        return cur;
    }
}

